tests/drivers/rtc/rtc_api/src/test_alarm.c

```diff
@@ -0,0 +1,143 @@
+
+		zassert_true((ret == 0) || (ret == -ENOTSUP),
+			     "Failed to clear and disable alarm callback");
+	}
+
+	/* Validate alarms supported fields */
+	for (uint16_t i = 0; i < alarms_count; i++) {
+		ret = rtc_alarm_get_supported_fields(rtc, i, &alarm_time_mask_supported);
+
+		zassert_true(ret == 0, "Failed to get supported alarm fields");
+
+		/* Skip test if alarm does not support the minute and hour fields */
+		if (((RTC_ALARM_TIME_MASK_MINUTE & alarm_time_mask_supported) == 0) ||
+		((RTC_TEST_ALARM_TIME_HOUR & alarm_time_mask_supported) == 0)) {
+			ztest_test_skip();
+		}
+	}
+
+	/* Set alarm time */
+	alarm_time_set.tm_min = RTC_TEST_ALARM_TIME_MINUTE;
+	alarm_time_set.tm_hour = RTC_TEST_ALARM_TIME_HOUR;
+	alarm_time_mask_set = (RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR);
+
+	for (uint16_t i = 0; i < alarms_count; i++) {
+		ret = rtc_alarm_set_time(rtc, i, alarm_time_mask_set, &alarm_time_set);
+
+		zassert_true(ret == 0, "Failed to set alarm time");
+	}
+
+	/* Validate alarm time */
+	for (uint16_t i = 0; i < alarms_count; i++) {
+		ret = rtc_alarm_get_time(rtc, i, &alarm_time_mask_get, &alarm_time_get);
+
+		zassert_true(ret == 0, "Failed to set alarm time");
+
+		zassert_true(alarm_time_mask_get == alarm_time_mask_set,
+			     "Incorrect alarm time mask");
+
+		zassert_true(alarm_time_get.tm_min == alarm_time_get.tm_min,
+			     "Incorrect alarm time minute field");
+
+		zassert_true(alarm_time_get.tm_hour == alarm_time_get.tm_hour,
+			     "Incorrect alarm time hour field");
+	}
+
+	/* Initialize RTC time to set */
+	timer_set = RTC_TEST_ALARM_SET_TIME;
+
+	gmtime_r(&timer_set, (struct tm *)(&time_set));
+
+	time_set.tm_isdst = -1;
+	time_set.tm_nsec = 0;
+
+	for (uint8_t i = 0; i < 2; i++) {
+		/* Set RTC time */
+		ret = rtc_set_time(rtc, &time_set);
+
+		zassert_true(ret == 0, "Failed to set time");
+
+		/* Clear alarm pending status */
+		for (uint16_t i = 0; i < alarms_count; i++) {
+			ret = rtc_alarm_is_pending(rtc, i);
+
+			zassert_true(ret > -1, "Failed to clear alarm pending status");
+		}
+
+		/* Wait before validating alarm pending status has not been set prematurely */
+		k_sleep(K_SECONDS(RTC_TEST_ALARM_TEST_NOT_PENDING_DELAY));
+
+		/* Validate alarm are not pending */
+		for (uint16_t i = 0; i < alarms_count; i++) {
+			ret = rtc_alarm_is_pending(rtc, i);
+
+			zassert_true(ret == 0, "Alarm should not be pending");
+		}
+
+		/* Wait for alarm to trigger */
+		k_sleep(K_SECONDS(RTC_TEST_ALARM_TEST_PENDING_DELAY));
+
+		/* Validate alarm is pending */
+		for (uint16_t i = 0; i < alarms_count; i++) {
+			ret = rtc_alarm_is_pending(rtc, i);
+
+			zassert_true(ret == 1, "Alarm should be pending");
+		}
+	}
+
+	/* Disable and clear alarms */
+	for (uint16_t i = 0; i < alarms_count; i++) {
+		ret = rtc_alarm_set_time(rtc, i, 0, NULL);
+
+		zassert_true(ret == 0, "Failed to disable alarm");
+
+		ret = rtc_alarm_is_pending(rtc, i);
+
+		zassert_true(ret > -1, "Failed to clear alarm pending state");
+	}
```

Comment:
Note - this code won't run if any of the previous `zassert` checks fail.  You should leave the alarm disable here since you're also verifying that the alarm cleared.  But you should still create an `after_fn` that also disables the alarms.  This assumes there's no side effect to disabling the alarm twice.

Probably best way to architect this is to create a test fixture in main.c
```
struct rtc_api_fixture {
    const struct device *rtc;
    alarms_count;
};

static void *rtc_api_setup(void)
{
	static struct rtc_api_fixture fixture;

	fixture.rtc = DEVICE_DT_GET(DT_ALIAS(rtc));
	fixture.alarms_count = DT_PROP(DT_ALIAS(rtc), alarms_count);

	return &fixture;
}

static void rtc_api_after(void *fixture)
{
	for (uint16_t i = 0; i < alarms_count; i++) {
		ret = rtc_alarm_set_time(rtc, i, 0, NULL);
	}
}

ZTEST_SUITE(rtc_api, NULL, NULL, NULL, rtc_api_after, NULL);
```

You'll then use `ZTEST_F()` for all your tests to pass in the fixture to each test.

Reply:
The reason I am not using the RTC API functions outside of unit tests is that I presume the RTC API calls are not working.

Each unit test is expecting that calling any RTC API function can fail and potentially hard fault. Calling said RTC API functions from within the before or after function is not safe for that reason.

before and after are IMO for performing safe context initialization, like clearing a k_event, or memsetting some buffers, things that are definitely not expected to fail :)

Reply:
Fair enough about not trusting the RTC API.  I would suggest adding the fixture to the RTC tests as a follow up PR to eliminate some duplicated code.

Reply:
I agree that's a good idea :)